contracts/reputation/Reputation.sol

```diff
@@ -14,81 +14,289 @@ pragma solidity >=0.5.0 <0.6.0;
+     */
+    function decrease(address _validator, uint256 _delta)
+        external
         onlyConsensus
+        isActive(_validator)
+        returns (uint256)
+    {
+        if (reputations[_validator] < _delta) {
+            reputations[_validator] = 0;
+        } else {
+            reputations[_validator] = reputations[_validator].sub(_delta);
+        }
+
+        return reputations[_validator];
+    }
+
+    /**
+     * @dev Function requires:
+     *          - only consensus can call
+     *          - a validator was not joined previously
+     *          - a validator address is not 0
+     *          - a withdrawal address is not 0
+     *          - a validator address is not same as its withdrawal address
+     *          - a validator approved in mOST token contract to transfer
+     *            a stake amount.
+     *          - a validator approved in wETH token contract to transfer
+     *            a stake amount.
+     */
+    function join(
+        address _validator,
+        address _withdrawalAddress
+    )
         external
-        view
+        onlyConsensus
     {
-        require(status[_validator] == ValidatorStatus.Undefined,
-            "No validator can rejoin.");
-        // continue
+        require(
+            _validator != address(0),
+            "Validator address is 0."
+        );
+
+        require(
+            _withdrawalAddress != address(0),
+            "Validator's withdrawal address is 0."
+        );
+
+        require(
+            _validator != _withdrawalAddress,
+            "Validator's address is the same as its withdrawal address."
+        );
+
+        require(
+            statuses[_validator] == ValidatorStatus.Undefined,
+            "No validator can rejoin."
+        );
+
+        require(
+            validators[_validator] == address(0),
+            "No validator can rejoin."
+        );
+
+        // adding a validator into the circular link-list of validators.
+        validators[_validator] = validators[SENTINEL_VALIDATORS];
+        validators[SENTINEL_VALIDATORS] = _validator;
+
+        statuses[_validator] = ValidatorStatus.Staked;
+        withdrawalAddresses[_validator] = _withdrawalAddress;
+        reputations[_validator] = initialReputation;
+
+        require(
+            mOST.transferFrom(_validator, address(this), stakeMOSTAmount),
+            "Failed to transfer mOST stake amount from a validator."
+        );
+
+        require(
+            wETH.transferFrom(_validator, address(this), stakeWETHAmount),
+            "Failed to transfer mOST stake amount from a validator."
+        );
     }
 
-    function logout(address _validator)
+    /**
+     * @dev Function requires:
+     *          - only consensus can call
+     *          - a validator has joined
+     */
+    function slash(address _validator)
+        external
         onlyConsensus
+        hasJoined(_validator)
+    {
+        statuses[_validator] = ValidatorStatus.Slashed;
```

Comment:
this is definitely incomplete; let's mark this as tobecontinued and `revert()`

Reply:
Done.